### PR TITLE
Add GraphQL entity type generation and streamline type system conversions

### LINKED ISSUE #97: Generate classes for each entity type
Rather than instantiating a generic Entity class and setting the fields individually like this
```
  let application = new Entity()
  application.setString('id', listingHash)
  application.setU256('expirationDate', listingResult.value0)
  application.setBoolean('whitelisted', listingResult.value1)
  application.setAddress('owner', listingResult.value2)
  application.setU256('deposit', listingResult.value3)
```
It would be better if we could instantiate an entity type specific class and set the data as a map like this
```
let application = new ApplicationEntity({
  id: listingHash,
  expirationDate: listingResult.value0,
  whitelisted: listingResult.value1,
  owner: listingResult.value2,
  deposit: listingResult.value3
})
```
Questions:
- How should we namespace these classes so they don't conflict?
- Should we change our current habit of injecting globals an instead make a module available so that people can manually import the classes they want at the top of the file?

### PR DESCRIPTION
**Pull request by Jannis**
Resolves #97. Resolves https://github.com/graphprotocol/tasks/issues/396. Depends on  https://github.com/graphprotocol/graph-ts/pull/14 and a new `graph-ts` release. 

1. All supported type conversions are now conveniently listed in https://github.com/graphprotocol/graph-cli/blob/jannis/graphql-types-v2/src/codegen/types/conversions.js, with type-related code generation being tested (partially) in https://github.com/graphprotocol/graph-cli/blob/jannis/graphql-types-v2/src/codegen/types/index.test.js.
2. In addition to improved/simplified type conversions, this PR introduces code generation for entity types defined in the GraphQL schema of subgraphs. An example can be found in https://github.com/graphprotocol/graph-cli/blob/jannis/graphql-types-v2/examples/example-event-handler/src/types/schema.ts.
3. The example subgraph now includes all supported field types in the GraphQL schema (https://github.com/graphprotocol/graph-cli/blob/jannis/graphql-types-v2/examples/example-event-handler/schema.graphql) and all supported Ethereum types and entity field getters/setters in the mapping (https://github.com/graphprotocol/graph-cli/blob/jannis/graphql-types-v2/examples/example-event-handler/src/mapping.ts). Building this subgraph in CI verifies that all APIs needed for using these types/getters/setters are available.

### CODE REVIEW

File: src/codegen/typescript.js [hunk 1]
```diff
@@ -0,0 +1,489 @@
+let immutable = require('immutable')
+let Map = immutable.Map
+
+//const schemaHasNamedType = (schema, name) =>
```
Comment (Jannis) [inline src/codegen/typescript.js]: Remove all this commented-out code.

File: src/codegen/typescript.js [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    this.type = type
+  }
+
+  toString() {
+    return `Value.${valueFromTypeFunction(this.type)}(${this.expr})`
+  }
+}
+
+const namedType = name => new NamedType(name)
+//const ethereumType = name => new EthereumType(name)
```
Comment (Jannis) [inline src/codegen/typescript.js]: Remove

File: src/codegen/typescript.js [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+const arrayType = name => new ArrayType(name)
+const param = (name, type) => new Param(name, type)
+const method = (name, params, returnType, body) =>
+  new Method(name, params, returnType, body)
+const staticMethod = (name, params, returnType, body) =>
+  new StaticMethod(name, params, returnType, body)
+const klass = (name, options) => new Class(name, options)
+const klassMember = (name, type) => new ClassMember(name, type)
+//const ethereumValueFromCoercion = (expr, type) =>
```
Comment (Jannis) [inline src/codegen/typescript.js]: Remove

File: src/codegen/typescript.js [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+//  new EthereumValueFromCoercion(expr, type)
+//const ethereumValueToCoercion = (expr, type) => new EthereumValueToCoercion(expr, type)
+const unionType = (...types) => new UnionType(types)
+const nullableType = type => new NullableType(type)
+const moduleImports = (nameOrNames, module) => new ModuleImports(nameOrNames, module)
+//const valueToCoercion = (expr, type) => new ValueToCoercion(expr, type)
```
Comment (Jannis) [inline src/codegen/typescript.js]: Remove

File: src/codegen/typescript.js [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+  staticMethod,
+  param,
+  //ethereumValueFromCoercion,
+  //ethereumValueToCoercion,
+  nullableType,
+  unionType,
+  moduleImports,
+  //valueToCoercion,
+  //valueFromCoercion,
+  //graphqlTypeToAssemblyScriptType,
```
Comment (Jannis) [inline src/codegen/typescript.js]: Remove the commented-out lines above.

**Review by leoyvens — COMMENTED** [review]
This is all good stuff, the type generation seems to be working wonders and we got a testing setup and validation improvements on the side, @Jannis strikes again!

I do have one comment. `conversions.js` is super clean and readable, but I think we cut it's line count in half. There's some duplication between `ETHEREUM_VALUE_TO_ASSEMBLYSCRIPT` and `ASSEMBLYSCRIPT_TO_ETHEREUM_VALUE`, same with `VALUE_TO_ASSEMBLYSCRIPT` and `ASSEMBLYSCRIPT_TO_VALUE`. I'd suggest merging them so there's only `ETHEREUM_VALUE_TO_ASSEMBLYSCRIPT` and `VALUE_TO_ASSEMBLYSCRIPT`, and for example and entry in `ETHEREUM_VALUE_TO_ASSEMBLYSCRIPT` would be:
```
['address', 'Address', to_code => `${to_code}.toAddress()`,
from_code: `EthereumValue.fromAddress(${from_code})]
```

**Comment by Jannis** [comment]
@leodasvacas There are a few subtle differences in the type conversion tables. For some conversions in either direction, the order matters (for matching). For other conversions, one direction uses a different matching pattern than the other direction.

I thought about merging the tables before but it's not trivial.

**Comment by Jannis** [comment]
The tests will fail until we make a new `graph-ts` release and bump its version in `examples/example-event-handler`.

**Comment by leodasvacas** [comment]
@Jannis I see, do you wanna add a comment noting that the tables could perhaps be merged but it's not so simple?

About the tests, do you wanna try landing the graph-ts PR first, and then using a git dependency on master so we can get them to pass?

**Comment by Jannis** [comment]
@leodasvacas Sounds like a plan, let's merge graph-ts, switch to graph-ts from master, get the tests to pass, then test graph-node as well and then make a release for graph-ts, merge this, make a new release for graph-cli, then merge graph-node. 😁 

I've added notes about the conversion tables in comments.

**Comment by leodasvacas** [comment]
@Jannis the rename caused a CI failure.

**Comment by Jannis** [comment]
Something is up with the commit Travis is building from for the PR — `allocate_memory` is gone in the PR.

**Comment by Jannis** [comment]
@leodasvacas The `export { allocate_memory }` statement in `global.ts` is breaking builds, both locally and in Travis. I'm going to revert it; we'll figure it out.

**Comment by Jannis** [comment]
Tests are passing, I tested the corresponding graph-node PR (https://github.com/graphprotocol/graph-node/pull/541) with 3-4 subgraphs and things are working!

I'm going to cut a new graph-ts release (bumping the minor version) and update the dependency in the example subgraph, so we can merge this as well.

**Comment by Jannis** [comment]
Let's get this over the finish line. Since @leodasvacas approved this prior to the `allocate_memory` revert, I'm going to go ahead and merge the PR.

**Pull request merged**